Sources/AWSLambdaRuntimeCore/Terminator.swift

```diff
@@ -0,0 +1,139 @@
+
+    private var storage: Storage
+
+    init() {
+        self.storage = Storage()
+    }
+
+    /// Register a shutdown handler with the terminator
+    ///
+    /// - parameters:
+    ///     - name: Display name for logging purposes
+    ///     - handler: The shutdown handler to call when terminating the Lambda.
+    ///             Shutdown handlers are called in the reverse order of being registered.
+    ///
+    /// - Returns: A `RegistrationKey` that can be used to de-register the handler when its no longer needed.
+    @discardableResult
+    public func register(name: String, handler: @escaping (EventLoop) -> EventLoopFuture<Void>) -> RegistrationKey {
+        let key = RegistrationKey()
+        self.storage.add(key: key, name: name, handler: handler)
+        return key
+    }
+
+    /// De-register a shutdown handler with the terminator
+    ///
+    /// - parameters:
+    ///     - key: A `RegistrationKey` obtained from calling the register API.
+    public func deregister(_ key: RegistrationKey) {
+        self.storage.remove(key)
+    }
+
+    /// Begin the termination cycle
+    /// Shutdown handlers are called in the reverse order of being registered.
+    ///
+    /// - parameters:
+    ///     - eventLoop: The `EventLoop` to run the termination on.
+    ///
+    /// - Returns: An `EventLoopFuture` with the result of the termination cycle.
+    internal func terminate(eventLoop: EventLoop) -> EventLoopFuture<Void> {
+        func terminate(_ iterator: IndexingIterator<[(name: String, handler: Handler)]>, errors: [Error], promise: EventLoopPromise<Void>) {
+            var iterator = iterator
+            guard let handler = iterator.next()?.handler else {
+                if errors.isEmpty {
+                    return promise.succeed(())
+                } else {
+                    return promise.fail(TerminationError(underlying: errors))
+                }
+            }
+            handler(eventLoop).whenComplete { result in
+                var errors = errors
+                if case .failure(let error) = result {
+                    errors.append(error)
+                }
+                return terminate(iterator, errors: errors, promise: promise)
+            }
+        }
+
+        // terminate in cascading, reverse order
+        let promise = eventLoop.makePromise(of: Void.self)
+        terminate(self.storage.handlers.reversed().makeIterator(), errors: [], promise: promise)
+        return promise.futureResult
+    }
+}
+
+extension LambdaTerminator {
+    /// Lambda terminator registration key.
+    public struct RegistrationKey: Hashable, CustomStringConvertible {
+        var value: String
+
+        init() {
+            // UUID basically
+            self.value = LambdaRequestID().uuidString
+        }
+
+        public var description: String {
+            self.value
+        }
+    }
+}
+
+extension LambdaTerminator {
+    private final class Storage {
```

Comment:
NIT: I don't really see the need for the Storage class here, since the TerminationHandler is a class as well. But I don't think that this should stop us from merging.

Reply:
I'd like to keep that abstraction